bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/AuditorElector.java

```diff
@@ -191,8 +192,14 @@ public void run() {
                         Thread.currentThread().interrupt();
                         submitShutdownTask();
                     } catch (Exception e) {
-                        LOG.error("Exception while performing auditor election", e);
-                        submitShutdownTask();
+                        if (e.getCause() instanceof KeeperException.ConnectionLossException
```

Comment:
It's better to retry with a delay time I think.

Reply:
IIRC ConnectionLoss is a non-recoverable exception.
To handle it properly, one has to re-create ZK client and deal with the watchers/cached state. Ephemeral nodes expire as result of such error too.
Re-initializing the client at random place when error happens is a rather big chunk of work IIRC and it easier to bounce the service in this case, hence https://github.com/apache/bookkeeper/pull/3374

Reply:
@dlg99 I saw it said connection loss is a recoverable exception.
```
Recoverable errors: the disconnected event, connection timed out, and the connection loss exception are examples of recoverable errors
```

And the code
```
       public static boolean isRecoverableException(int rc) {
            return KeeperException.Code.CONNECTIONLOSS.intValue() == rc
                    || KeeperException.Code.OPERATIONTIMEOUT.intValue() == rc
                    || KeeperException.Code.SESSIONMOVED.intValue() == rc
                    || KeeperException.Code.SESSIONEXPIRED.intValue() == rc;
        }
```